db/core/htrie.c

```diff
@@ -792,22 +977,158 @@ tdb_htrie_insert(TdbHdr *dbh, unsigned long key, void *data, size_t *len)
+			break;						\
+		body;							\
+		rec = (TdbRec *)((char *)rec + rlen);			\
+	} while ((char *)rec + sizeof(*rec) - (char *)bckt		\
+		 <= TDB_HTRIE_MINDREC)
+
+int
+tdb_htrie_remove(TdbHdr *dbh, unsigned long key, tdb_eq_cb_t *eq_cb, void *data,
+		 tdb_before_remove_cb_t *bf_remove_cb, bool force)
+{
+	int bits = 0;
+	TdbRec *rec;
+	unsigned long o;
+	TdbBucket *bckt, *prev, *b_tmp;
+	TdbHtrieNode *node = TDB_HTRIE_ROOT(dbh);
+	int r = -ENOENT, has_alive;
+
+	o = tdb_htrie_descend(dbh, &node, key, &bits);
+	if (!o)
+		return r;
+
+	prev = bckt = TDB_PTR(dbh, o);
+	BUG_ON(!bckt);
+
+	TDB_DBG("Remove htrie record for key=%#lx force=%d bucket=[%p]", key,
+		force, bckt);
+
+	write_lock_bh(&bckt->lock);
+	/*
+	 * Iterate all records in the head bucket. Try to free suitable
+	 * records. The head bucket can't be reclaimed, but other buckets
+	 * must be reclaimed if no live records in it.
+	 */
+	if (tdb_bucket_is_complete(bckt) || force) {
+		rec = TDB_HTRIE_BCKT_1ST_REC(bckt);
+		TDB_REMOVE_FOREACH_REC({
+			if (tdb_live_rec(dbh, rec)
+			    && tdb_eval_eq_cb(eq_cb, rec, data))
+			{
+				tdb_eval_bf_remove_cb(bf_remove_cb, rec);
+				tdb_htrie_free_rec(dbh, rec);
+				r = 0;
+			}
+		});
+	}
+
+	if (!bckt->coll_next) {
+		write_unlock_bh(&bckt->lock);
+		return r;
+	}
+
+	bckt = TDB_HTRIE_BUCKET_NEXT(dbh, bckt);
+
+	write_lock_bh(&bckt->lock);
+	/* Iterate all buckets except the head bucket. */
+	while (bckt) {
+		if (tdb_bucket_is_complete(bckt) || force) {
+			has_alive = false;
+			rec = TDB_HTRIE_BCKT_1ST_REC(bckt);
+			TDB_REMOVE_FOREACH_REC({
+				if (tdb_live_rec(dbh, rec)) {
+					if (tdb_eval_eq_cb(eq_cb, rec, data)) {
+						tdb_eval_bf_remove_cb(bf_remove_cb,
+								      rec);
+						tdb_htrie_free_rec(dbh, rec);
+						r = 0;
+					} else {
+						has_alive = true;
+					}
+				}
+			});
+
+			/*
+			 * Once bucket doesn't have any live records, reclaim it.
+			 */
+			if (!has_alive) {
+				prev->coll_next = bckt->coll_next;
+				b_tmp = bckt;
+				bckt = TDB_HTRIE_BUCKET_NEXT(dbh, bckt);
+				write_unlock_bh(&b_tmp->lock);
+				/*
+				 * Reclaim bucket's memory. Do reclaim after
+				 * unlocking relying on previous bucket lock.
+				 */
+				tdb_put_blk(dbh, TDB_OFF(dbh, b_tmp));
+				if (bckt)
+					write_lock_bh(&bckt->lock);
+				else
+					write_unlock_bh(&prev->lock);
+				continue;
+			}
+		}
+
+		if (!bckt->coll_next) {
+			write_unlock_bh(&bckt->lock);
+			write_unlock_bh(&prev->lock);
+			break;
+		}
+
+		b_tmp = bckt;
+		bckt = TDB_HTRIE_BUCKET_NEXT(dbh, bckt);
+		write_lock_bh(&bckt->lock);
+		write_unlock_bh(&prev->lock);
+		prev = b_tmp;
+	}
+
+	return r;
+}
+
 TdbBucket *
 tdb_htrie_lookup(TdbHdr *dbh, unsigned long key)
 {
 	int bits = 0;
 	unsigned long o;
+	TdbBucket *b;
 	TdbHtrieNode *root = TDB_HTRIE_ROOT(dbh);
 
 	o = tdb_htrie_descend(dbh, &root, key, &bits);
 	if (!o)
 		return NULL;
+	b = TDB_PTR(dbh, o);
+	read_lock_bh(&(b)->lock);
+	if (!tdb_bucket_is_complete(b)) {
```

Comment:
I don't think we need to implement #500 right now, but there is a comment https://github.com/tempesta-tech/tempesta/issues/500#issuecomment-377051964 , which is aligned with `incomplete` buckets (why not records?), so it'd be good to write a good `TODO #500` comment and align the current implementation with the upcoming task.